File: src/dotnet/Ansight.Tools.SecureStorage/GetSecureStorageValueTool.cs
namespace Ansight.Tools.SecureStorage;

public sealed class GetSecureStorageValueTool : ITool
{
    private readonly SecureStorageToolsOptions options;

    public GetSecureStorageValueTool(SecureStorageToolsOptions? options = null)
    {
        this.options = options ?? SecureStorageToolsOptions.Default;
    }

    public string Category => "secure";

    public ToolScope Scope => ToolScope.Read;

    public string Id => SecureStorageToolIds.GetValue;

    public string Name => "Get Secure Storage Value";

    public string Description => "Reads a decrypted value from the configured secure storage backend.";

    public string Keywords => "secure storage keychain keystore encrypted get";

    public ToolSchema ArgumentsSchema => SecureStorageToolSchemas.GetValueArguments;

    public ToolSchema ResultSchema => SecureStorageToolSchemas.GetValueResult;

    public ToolSecurity Security => SecureStorageToolSecurityProfiles.GetValue;

    public Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            return Task.FromResult(ToolResult.Success(SecureStorageSupport.GetValue(options, arguments)));
        }
        catch (PlatformNotSupportedException exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "secure_platform_unsupported"));
        }
        catch (Exception exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "secure_get_failed"));
        }
    }
}

File: src/ios/Sources/AnsightToolsSecureStorage/GetSecureStorageValueTool.swift
import AnsightCore
import Foundation

public final class GetSecureStorageValueTool: AnsightTool {
    private let options: AnsightSecureStorageToolsOptions

    public init(options: AnsightSecureStorageToolsOptions = .default) {
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightSecureStorageToolIds.getValue,
            name: "Get Secure Storage Value",
            description: "Reads a decrypted value from the configured secure storage backend.",
            category: "secure",
            policy: .critical,
            keywords: "secure storage keychain keystore encrypted get",
            argumentsSchema: AnsightSecureStorageToolSchemas.getValueArguments,
            resultSchema: AnsightSecureStorageToolSchemas.getValueResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        do {
            return .success(try AnsightSecureStorageSupport.getValue(options: options, arguments: arguments))
        } catch AnsightSecureStorageToolError.platformUnsupported(let message) {
            return .failure(message, errorCode: "secure_platform_unsupported")
        } catch {
            return .failure(error.localizedDescription, errorCode: "secure_get_failed")
        }
    }
}
